### DISCUSSION TITLE
Design of shared connection configuration for EdgeDB clients

### DISCUSSION THREAD

**Discussion by elprans** [geldata/gel#1683] [General]
This is a continuation of the discussion started in edgedb/edgedb-cli#91.

The goal is to provide a mechanism that makes configuration of basic connection parameters secure and convenient by default.  

edgedb/edgedb-cli#124 implements support for the generation of a `credentials` file in a well-known location for each EdgeDB instance during `edgedb server init`.  The CLI itself can then use the file by default when connecting to the specified instance (via the `-I` flag).  Naturally, the CLI is not the only way to interact with an EdgeDB instance, so we need a similar mechanism for language bindings.

I propose adding the new `instance` parmeter to `connect()`.  It can be a string, in which case it signifies a locally-configured instance and uses the appropriate `credentials` file generated by the CLI, or it can be an object that represents the instance address and credentials.

For example:

```python
# connect to the "myapp" instance, take
# information from the known credentials file
conn = edgedb.connect(instance="myapp")
```

or

```python
instance = edgedb.Instance(host='...', port='...', password='...')
conn = edgedb.connect(instance=instance)
```

There can be multiple ways to construct an `Instance` object:

```python
instance = edgedb.Instance.from_credentials_file(...)
# override the default database
instance.database = 'foo'
```

or

```python
# take connection information from well-known environment variables
instance = edgedb.Instance.from_environment()
# optionally override instance parameters
```

The current keyword-argument based API will be deprecated in favor of the above.  The advantage of a separate `Instance` configuration object is that we don't need to repeat ourselves in `create_pool` and other future connection interfaces.

**Comment by 1st1** [comment]
+1 overall.

Few things I want to propose:

* Make the `instance` parameter of the `connect()` function positional-only. So it becomes `connect('myapp')`. This has the benefit of saving extra keystrokes and will nicely translate to edgedb-js.  This way we can also accept DSNs.

* Rename `edgedb.Instance` to `edgedb.InstanceConnector`.

* Our CLI tools and Python/JS/etc APIs should work the same. If we have a default instance that CLI uses by default our programmatic APIs should do the same. So `edgedb.connect()` would try to connect to the default instance.

* Because `edgedb.connect()` would no longer try to connect to the port 5656 if all other option fail, our CLI should work the same. E.g. CLI tries to connect to the default instance first; if there's no default instance - prompt user to either specify an instance name or specify ports.

**Comment by tailhook** [comment]
I have a bunch of thoughts:

1. I don't like name "instance". It's clear what it is in the context of `server` command, but I think it would make me google or "what is instance in edgedb" in the first place (i.e. is it some abstraction like "database" and postgres "schema", or is it a separate cluster, or is it edgedb-cloud-specific thing?)
2. I like `Xxx.from_environment()` much more than implicitly using `EDGEB_HOST` and friends
3. I don't like `instance = Xxx.from_credentials(); instance.database = 'yy'` I think `connect(credentials=..., database='xxx')` is a much nicer way to override database
4. @1st1 has mentioned in zoom call that the biggest problem with keyword arguments is implicit "and" between parameters. But actually this the same with builder pattern: every method is expected to be orthogonal to any other, except when documented otherwise. So while `builder.credentials(dict)` and `builder.credentials_file(path)` are easy to correlate, `builder.set_hosts(...)` and `builder.failover_timeout(..)` are not so much. So it's mostly two things about builders that make difference (in dynamic languages): constructor and the actual subclass.
5. It's still not clear: @elprans, do you propose to drop all other arguments to connect? 
6. I'm kinda sceptical if `connect('my_instance')` automatically supports DSN `connect('edgedb://...')` by matching string. @1st1, is it what you propose?

So my proposals so far:

# Option 1

Name what @elprans proposed a `edgedb.Config` rather than `Instance`, and accept it as a first argument instead of dsn:

1. Current semantics would be `connect(Config.from_environment())`
2. Dsn support isn't that hard: `connect(dsn="edgedb://...")` which is equivalent of `connect(Config.from_dsn("edgedb://..."))`
3. Named instances are `connect('name')` which is equivalent to `connect(Config.named('name'))`
4. Credentials file`connect(Config.from_credentials_file ('my.file.json'))`
5. Move all arguments from `connect` to Config: `connect(Config(host='localhost', port=1234))`
6. Use just user/file from file: `connect(Config(host='localhost', port=1243).with_credentials_file('my_file.json'))`

The problem with this option is that we just moved everything to `Config` and that is still huge uber-object with lots of things. I.e. do we want connection pool settings to be contained in config? Or arey they still keyword arguments? Or do we have `PoolConfig` builder, which has `PoolConfig.defaults().with_connection_config(Config.from_environment())`?

# Option 2

So option 2 is to keep current semantics, but add three parameters to connect:
1.  `connect(local_instance='abc')` this uses everything from credentials file, ensuring that it's a local instance (in some cases this still connects to a remote host, this is needed for VM in windows, so this is still considered local)
2. `connect(host="production-host.example.org", crendentials_file="/volumes/creds/edgedb.json")` -- uses just user, password and database (latter is overridable in the params of the connect). We don't need named credentials here as I still expect no user home directory on production.
3. `connect(cluster_config=ClusterConfig(..), credentials_file="/volumes/creds/edgedb.json")` real config with high availability (i.e. renamed `ha_config` [from this proposal](https://github.com/edgedb/edgedb-rust/issues/45#issuecomment-675772132))

`connect('edgedb://..'` is left intact for users that want that.

I think this keeps simple ways to connect simpler and similar to other databases. And splits credentials from network/cluster/HA config in quite elegant way, without any extra nesting (i.e. no `Config.with_cluster_config(...)`)

**Reply by elprans** [reply]
> 5. It's still not clear: @elprans, do you propose to drop all other arguments to connect?

Yes. Deprecate and eventually remove.  Having _both_ a fluent connection config builder, _and_ redundant keyword arguments in connection methods is messy and confusing, in my opinion.  The only useful exception might be the database name.

> Name what @elprans proposed a `edgedb.Config`

`Config` can be confused with server-side configuration, and we are talking about purely a client connection configuration, so we should either qualify it as `ClientConfig` or use another name, like `Connector` (more on that later).

> I.e. do we want connection pool settings to be contained in config

No, we don't want that.  The idea here is that we encapsulate the configuration and logic of establishing *one* connection in a `Connector` object.  Pool configuration is orthogonal to that and should be passed separately.

In this approach it's easy to actually delegate the creation of a data stream to the `Connector` object.  This way `connect()` or `create_pool()` don't have to try to discern if a Unix socket or TCP, or 5 layers of TLS are used to pass protocol messages.

> So option 2 is to keep current semantics, but add three parameters to connect.

Adding parameters to `connect()` doesn't really scale very well.  `credentials_file` is, for sure, not going to be the only way to supply credentials (think secrets service or a certificate manager), and so this split is rather arbitrary and focuses on just one use case. 

> I think this keeps simple ways to connect simpler

`connect('myapp')` is as simple as it gets.  The homedir shouldn't be the _only_ place where instance credentials are looked up (as we have system-level instances), and so advising users to mount instance credentials to `/var/lib/edgedb/credentials` (or wherever we keep them) is not that complicated.  Other than that, you always have an option to pass the config either as DSN or via environment.

> and similar to other databases

The common denominator among database connectors is usually some support for DSN strings, and we support that.  Aside from that, I don't see any value in retaining the "100 keyword arguments in connect()" approach.

An important benefit of a single connection configuration object is that you can introspect it, dump it, log it etc.  A tangle of objects and method keywords don't have that clarity.

**Reply by 1st1** [reply]
+1 to what Elvis said.

@tailhook 
> I'm kinda sceptical if connect('my_instance') automatically supports DSN connect('edgedb://...') by matching string. @1st1, is it what you propose?

Yes, `edgedb://` is unlikely to be an instance name, we can even ban names like that.

> It's still not clear: @elprans, do you propose to drop all other arguments to connect?

Yes. We might have arguments to customize connections (custom object factories or middleware), but the connection configuration is going to be passed as one arg.

**Comment by tailhook** [comment]
Okay, so I'm doing single positional arg which is DNS-or-name and postpone everything else to the next release. This is a little bit awkward in JS as you need `connect(undefined, {host,port})` if you want to specify arguments the old way. But this should be solved when proper builders are there.

**Reply by 1st1** [reply]
> Okay, so I'm doing single positional arg which is DNS-or-name and postpone everything else to the next release. 

Perhaps I confused you here -- we'll do the positional-only DNS-or-name-or-instance-object later, but for now I think we can keep the current API working.

**Comment by agritheory** [comment]
How would one specify sync/async/pool connections in this API? Is that part of the credentials file?

@elprans example looks like sync python (clear, executable pseudo-code; not a criticism)
```python
# connect to the "myapp" instance, take
# information from the known credentials file
conn = edgedb.connect(instance="myapp")
```
So to approximate the other options (at least in python)
```python
conn = await edgedb.connect_async(instance="myapp")
```
And 
```python
conn = await edgedb.connect_async_pool(instance="myapp")
```
Connecting from Node doesn't seem like it would have the same API issue, I can't speak to Rust. 

Since naming things is hard, I'd like to propose 'Connection' instead of `Instance` or `InstanceConnection` or `ConnectionInstance`. It seems reasonable that `edgedb.connect` could return a wrapper that encapsulates the credentials and method.

**Reply by 1st1** [reply]
You got it right, except that `instance` is going to be a positional-only parameter. So it'll look like `edgedb.connect('myapp')` in Python for sync and in JS, and `edgedb.connect_async('myapp')` in Python for async.

**Reply by 1st1** [reply]
> Since naming things is hard, I'd like to propose 'Connection' instead of Instance

`Connection` would also be confusing, since this isn't really a connection object but a "connection configuration" object. That said, I also find `Instance` confusing, so perhaps something like `Connector` would make more sense.

**Comment by tailhook** [comment]
Just noticed, that by restricting `connect("name", host="some-hostname")` we make troubles for dockerized installations.

I.e. `edgedb` CLI should be able to connect to the database using localhost. But containers need a different hostname (that is also not accessible from host system). This is obviously, solved by `connect(ConnectConfig.named("name").override_hostname('some-host'))`, but this is quite big departure for quite a simple case.

**Reply by elprans** [reply]
> we make troubles for dockerized installations.

Do you mean the server running in a container, or client app running in a container?

**Reply by tailhook** [reply]
I mean when both server and app are in container, but CLI is not (which sounds as quite viable setup).

**Comment by fmoor** [comment]
This is what I have understood the direction of this discussion to be. Am I on the right track?

- The pool options type is not defined here but should follow a similar pattern as the Connector type.
- setters replace the current value.

# python
```python
def connect(connector: Connector, /) -> BlockingIOConnection: ...
def connect_async(connector: Connector, /) -> AsyncIOConnection: ...
def create_async_pool(connector: Connector, /, pool_options: Optional[PoolOptions]) -> AsyncIOPool: ...

class Connector:
    @classmethod
    def from_dsn(cls, dsn: str) -> Connector: ...

    @classmethod
    def from_file(cls, instance_name_or_arbitrary_path: str) -> Connector: ...

    @classmethod
    def from_env(cls) -> Connector: ...

    def __init__(self, addrs: List[Address], user: str, database: str,
                 password: Optional[str],
                 connect_timeout: Optional[datetime.timedelta],
                 command_timeout: Optional[datetime.timedelta],
                 wait_until_available: Optional[datetime.timedelta],
                 server_settings: Optional[Map[str, str]])

    def set_addrs(self, addrs: List[Address]) -> Connector: ...
    def set_user(self, user: str) -> Connector: ...
    def set_password(self, password: str) -> Connector: ...
    def set_database(self, database: str) -> Connector: ...
    def set_connect_timeout(self, timeout: datetime.timedelta) -> Connector: ...
    def set_command_timeout(self, timeout: datetime.timedelta) -> Connector: ...
    def set_wait_until_available(self, timeout: datetime.timedelta) -> Connector: ...
    def set_server_settings(self, settings: Map[str, str]) -> Connector: ...
```

# js
```javascript
function connect(connector: Connector): Promise<Connection>
function createPool(connector: Connector, options: PoolOptions | null): Promise<Pool>

class Connector {
	static fromDSN(dsn: string): Connector
	static fromFile(name: string): Connector
	static fromEnv(): Connector

	constructor(
		addrs: Array<Address>,
		user: string,
		database: string,
		password: string | null,
		connectTimeout: Number | null,
		commandTimeout: Number | null,
		waitUntilAvailable: Number | null,
		serverSettings: { [key: string]: string },
	)

	setAddrs(addrs: Array<Address>): Connector
	setUser(user: string): Connector
	setPassword(pass: string): Connector
	setDatabase(db: string): Connector
	setConnectTimeout(timeout: Number): Connector
	setCommandTImeout(timeout: Number): Connector
	setWaitUntilAvailable(timeout: Number): Connector
	setServerSettings(self, settings: { [key: string]: string }): Connector
}
```

# go
```go
func Connect(*Connector, *PoolOptions) (Pool, error)
func ConnectOne(*Connector) (Conn, error)

func DSN(string) *Connector
func File(string) *Connector
func Env() *Connector
func NewConnector(
	addrs []Address,
	user, password, database string,
	connectTimeout, commandTimeout, waitUntilAvailable time.Duration,
	serverSettings map[string]string,
) *Connector

type Connector struct {
	// all fields are private
}

// Err returns an error if the configuration is not valid.
func (c *Connector) Err() error

func (c *Connector) Addrs([]Address) *Connector
func (c *Connector) User(string) *Connector
func (c *Connector) Password(string) *Connector
func (c *Connector) Database(string) *Connector
func (c *Connector) ConnectTimeout(time.Duration) *Connector
func (c *Connector) CommandTimeout(time.Duration) *Connector
func (c *Connector) waitUntilAvailable(time.Duration) *Connector
func (c *Connector) ServerSettings(map[string]string) *Connector
```

**Reply by tailhook** [reply]
Looks like yes. Except I think we don't have to put "server settings" to a public API yet.

**Reply by fmoor** [reply]
how should transaction & retry options fit into this? Do you have to have a connection/pool before you can set transaction or retry options using a `.with_xxx_options()` method?

**Reply by tailhook** [reply]
> how should transaction & retry options fit into this? Do you have to have a connection/pool before you can set transaction or retry options using a `.with_xxx_options()` method?

Yes. Options described in this discussion are used to establish connection/pool. And `with_xxx` can be used at any moment and different options can be applied to a different transactions (even at the same time in the case of connection pool). So while we could allow setting defaults before making a connection we don't need it (so don't bother duplicating API).

**Comment by elprans** [comment]
(reposting this, as the thread got mixed up somehow)

In light of #2396 (`edgedb init-project`) and further thinking about this, I propose the following amendments to the discussion above:

1. Keep the current behavior of always reading from environment.

   The reason here is that the app code shouldn't be affected by how the app is deployed and tested, and passing configuration via environment is a fundamental practice, especially where containers are involved.  Instead, we should add a knob to explicitly *opt out* of reading environment variables.  Something like `Instance.from_scratch()`.
   
2. As also proposed in #2396, the behavior of bare `connect()` is as follows:

   ```python
   if have_env():
       config = Instance.from_env()
   elif exists(project_info):
       config = Instance.from_project_info()
   else:
       raise

   if 'EDGEDB_DATABASE' in os.environ:
       config.set_database(os.environ['EDGEDB_DATABASE'])
   ```
   
   Where `from_env()` is defined as:
   
   ```python
   if 'EDGEDB_DSN' in os.environ:
       config = Instance.from_dsn(environ['EDGEDB_DSN'])
   elif 'EDGEDB_INSTANCE' in os.environ:
       config = Instance.from_credentials_file(
           find_credentials(environ['EDGEDB_INSTANCE']),
       )
   elif 'EDGEDB_CREDENTIALS_FILE' in os.environ:
       config = Instance.from_credentials_file(
           find_credentials(environ['EDGEDB_INSTANCE']),
       )
   elif 'EDGEDB_HOST' in os.environ or 'EDGEDB_PORT' in os.environ:
       config = Instance(
          host=os.environ['EDGEDB_HOST'],
          ...,
       )
   ```
   
   The `_DSN`, `_INSTANCE`, `_CREDENTIALS_FILE` and granular (`_HOST`, `_PORT`) variables are mutually exclusive, i.e. cannot be set at the same time.  This is to prevent errors and to avoid assigning priority to environment variables.
   
3. The `connect()` should accept the `database` parameter directly.  The rationale is that while the above methods allow for flexible addressing of a given EdgeDB *instance*, the app might want to store its data in multiple databases.  For example, an app consisting of multiple services.  In production each service might have its own EdgeDB instance, but for a dev or test deployment it would make sense to run everything on the same instance.  Example:

   ```python
   # service1.py
   conn = edgedb.connect(database='service1')
   ...
   # service2.py
   conn = edgedb.connect(database='service2')
   ...
   ```

**Reply by elprans** [reply]
From @tailhook:

Generally agree. But I'm not sure I understand the first point.

Does `Instance()` use environ? Does `Instance(host='edgedb')` use environ? How do you do `Instance(**config)`? Does it look like this?:
```
inst = Instance.from_scratch()
inst.set_host(config['host'])
inst.set_port(config['port'])
```
Which isn't very pythonic. This makes the following even more messy:
```
inst1 = Instance(**config['db1'])
inst2 = Instance(**config['db2'])
```

Also even in simpler examples:
```
inst = Instance.from_env()
inst.set_host("other-host")
```
Makes much more clear idea that you are overriding some thinkgs in the environment, than:
```
inst = Instance()
inst.set_host("other_host")
```
Which reads as you only configured host.

So I would keep `.connect()` and `.connect(database='')` as the shortest form of using environment. And `Instance.from_env()` if you need special case.

**Reply by elprans** [reply]
> So I would keep `.connect()` and `.connect(database='')` as the shortest form of using environment. And `Instance.from_env()` if you need special case.

I'm good with this.
